cdc_shell.c

```diff
@@ -329,6 +353,350 @@ static void cdc_shell_cmd_uart(int argc, char *argv[]) {
     }
 }
 
+#define cdc_shell_msg_pin(pinn, fmt, ...) cdc_shell_msg("%6s [%d]: " fmt,      \
+                    gpion_to_str(pinn), gpion_pin_get_free(pinn, 0), ## __VA_ARGS__)
+
+#define cdc_shell_gpion_pin_ref(n) BUILD_BUG_ON_RET(                           \
+    ((gpion_pin_t)(gpio_pin_last + n) <= gpio_pin_last),                       \
+    (gpion_pin_t)(gpio_pin_last + n))
+
+#define cdc_shell_gpion_pin_ref_all      cdc_shell_gpion_pin_ref(1)
+#define cdc_shell_gpion_pin_ref_free     cdc_shell_gpion_pin_ref(2)
+#define cdc_shell_gpion_pin_ref_blocked  cdc_shell_gpion_pin_ref(3)
+#define cdc_shell_gpion_pin_ref_occupied cdc_shell_gpion_pin_ref(4)
+#define cdc_shell_gpion_pin_ref_uart1    cdc_shell_gpion_pin_ref(5)
+#define cdc_shell_gpion_pin_ref_uart2    cdc_shell_gpion_pin_ref(6)
+#define cdc_shell_gpion_pin_ref_uart3    cdc_shell_gpion_pin_ref(7)
+#define cdc_shell_gpion_pin_ref_uart     cdc_shell_gpion_pin_ref_uart1
+
+#define cdc_shell_pin_extra_func(func, var, pred, name, ...) do {              \
```

Comment:
Here is not clearly understandable meta-magic, should place extra comments for it and maybe move to external file?

Reply:
Added comments. Think about split cdc_shell.c file splitting on sub-systems like next:

     cdc_shell/general.c
     cdc_shell/uart.c
     cdc_shell/gpio.c
     cdc_shell/config.c

What does @r2axz think about it?